### ISSUE TITLE
Cross board size learning. Network topology.

### ISSUE THREAD

**Issue by ThomasCabaret** [gcp/leela-zero#907]
Hi, I was wondering why it was so easy for human to use something learned on a small board on a larger board and so difficult for AZ/LZ to do the same. I came up with this:
With the current AZ/LZ neural topology the side effects (I mean the impacts of the board sides) are embedded into the cnx weights and based on the topological location of the output neuron in it's 19x19 layer, but embedding those side effects is not the only way to proceed, we can actually force the network to COMPUTE them.
For that the idea would be to add one input layer to express the fact that a given intersection is part of the grid or not: SIDES ARE INPUTS! (and we work with a global input size larger than the max board size we want to cover)
Then, for a given channel on a given layer, we force the filter (the 3x3 cnx weights) to be
shared for any of the 361 computations of this channel on this layer. I mean that instead of having 19x19x3x3xNbChannel independent cnx weights by layer you get rid of the 19x19x due to this sharing (but still the number of neurons by layer is unchanged)
The required difference of behavior (because of course we do not want the same "production" close to the sides) would then come from the use of the "side/out of board" inputs. Cnx weights dont "know" the side effect any more, but they know how to compute it. What we have done is forcing the translation invariance in the network computation, that way the sides/effect impact has to be computed from the input only (as "sides" here are inputs)
Yes we can be afraid of the low level of expression this new topology has but it can be balanced by:
-A higher number of channels by layer (allowing the network to express more rich pieces of information to give to the next layer, mostly to give him the power to describe the side effect he "sees")
-A deeper network.
-Larger convolutionnal kernels (5x5 instead of 3x3 etc...)
As this architecture use barely 361 less independent cnx, we have room for increasing other stuffs.

-You could argue that by example on the first layer, except directly on the side the computation(function) done by the network would be identical and claim that is bad. But is it really the job of a first layer to influence the computation based on it's location? It's quite easy to see that in go, there are a lot of microscopic unbiased useful concepts you can build that are mostly invariant, by example recognizing strings of stones counting liberties etc... That is also true in more macroscopic layers: a network able to recognize or value a dead group somewhere would/should be able to recognize or value it anywhere. (and compare/choose/balance base on the context would/should be the role of the next layers only...)

-Most importantly the original aim of all of this: training on small board for use on larger ones. Any training on a board, will generate UNBIASED (by the global board size) microscopic concepts/value computations, all of this is a huge gain in the training on a larger board size that can/will reuse those already converged local computations.

Last layer/cross board training:
A good thing would be to maximize the "cross board" ability of this network and for that
the last layers should performed a task similar in essence between all board sizes. I base my guess on my Go knowledge, at the end, what we basically do, is comparisons of values, even if they are more rich than scalar values (multi channel), we compare them together and to whole board to conclude. So I suggest this last layers topology:
I will refer as direct-parents all the previous layer channel output at the same location than the considered channel (on their 19x19 layer map), and other-parents those at a different locations.  We can do a full connected layer but forcing/sharing the same cnx weights between cnx to "other-parents" (see that as an atopological comparison between the local value and other values, with all other values playing the same role) we can keep a multi channel output for this layer and then reduced it to a scalar in a last next layer....

For cross-board learning: we can add convolutional layers just before this last one, initialized
to identity. (you train it on 8x8 by example, you add some room/new layers just before the last layers initialized to identity then train on 9x9 and so on). RQ: for any board size training the input size is 19x19 it's just the input map indicating the sides that changes.

In a nutshell: really fewer cnx using the translational invariance of the game, balanced by a higher number of channels, larger depth or convolutional kernels size, seems to allow a more board-size-independent-computation that can be more easily trained from small boards to larger ones in hope of a really faster convergence.

Feel free to shared you thoughts about this.
Regards,

**Comment by gcp** [comment]
The board sides *ARE* part of the network input, as a consequence of there being a side to move plane, which interacts with how convolutional filters work.

In the original AGZ paper, only black got such a plane, so black is told where the sides are, but white is not. This asymmetry was removed in Leela Zero. This is even described in the README.

**Comment by ThomasCabaret** [comment]
The idea is not only to put sides as inputs but to combine it with forcing the translationnal invariance by sharing connexion weights on convolutionnal filters. Combinations of the two (invariance and sides input) allow any subcomputation on a part of the board to be independant of the context in which this part is included (letting the next layers to deal with the context). This way, a lot of thing learned by a network on a small board can be reused on a larger one.

**Comment by alreadydone** [comment]
I found this post when I checked my unread mails today; it must be too long that I didn't read it immediately. The 3x3 and 1x1 convolutional weights are indeed shared among different locations on the board; I don't know any convolutional layer that doesn't work this way. I think cross-board generalization and transfer of knowledge are possible by freezing all but the last few layers (including the value and policy heads which have fully connected layers whose weights are not shared among different locations) and feeding in training data generated on different boards. Similar idea appeared in https://github.com/gcp/leela-zero/issues/753

**Comment by Eddh** [comment]
It should also be possible to randomly set the board size for each training game, to 9x9, 13x13 or 19x19. But that's something for later. Hard to know if it would be beneficial.
